src/firewall/core/rich.py

```diff
@@ -13,7 +13,8 @@
 
 
 class _Rich_Entry:
-    pass
+    def check(self, family=None):
+        pass
```

Comment:
I don't think we need to pass `family` here. AFAICS, the only `check()` using it is `Rich_Reject`, but I don't see any code actually passing family to the constructor.

Reply:
Various "entry" types honor the family during check. For example, `Rich_Source.check()`, `Rich_Destination.check()`, `Rich_ForwardPort.check()`, `Rich_Reject.check()`. 

The relevant part is, while not all of them honor a family, some do. Consequently, for `_Rich_Entry.check()` the family must be an argument, to perform the necessary validation.

> but I don't see any code actually passing family to the constructor.

"constructor"? The `_Rich_Entry` instances don't actually have an address family property. Instead, the `Rich_Rule` has a family, and it passes it down to the `check()` method.

Reply:
> "constructor"? The _Rich_Entry instances don't actually have an address family property. Instead, the Rich_Rule has a family, and it passes it down to the check() method.

Okay. I think we can improve this.

If `family` is important to the element, e.g. `Rich_Source`, then it should be known to the object, i.e. `self.family`. Or we need a reference to the containing `Rich_Rule` so the element knows it's own context.

Many of the validation checks are "was family specified?". Only a few do real checks based on the family.

e.g.
```
            if family is None:
                raise FirewallError(errors.INVALID_FAMILY)
```

Reply:
Another example of why this would be useful.

This code is in `Rich_Rule` instead of `Rich_Audit` because it needs `self.action`.

```
        if self.audit is not None:
            if type(self.action) not in [Rich_Accept, Rich_Reject, Rich_Drop]:
                raise FirewallError(errors.INVALID_AUDIT_TYPE, type(self.action))
```

If we had a reference to the containing `Rich_Rule` we could have this code is `Rich_Audit.check()`.

```
        if type(self.parent.action) not in [Rich_Accept, Rich_Reject, Rich_Drop]:
                raise FirewallError(errors.INVALID_AUDIT_TYPE, type(self.parent.action))
```

`parent` may not be the best name.